compiler-core/src/format.rs

```diff
@@ -1578,20 +1623,48 @@ where
 
 pub fn comments_before<'a>(
     comments: &'a [Comment<'a>],
+    empty_lines: &'a [usize],
     limit: usize,
-) -> (impl Iterator<Item = &'a str>, &'a [Comment<'a>]) {
-    let end = comments
+    retain_empty_lines: bool,
+) -> (
+    impl Iterator<Item = Option<&'a str>>,
+    &'a [Comment<'a>],
+    &'a [usize],
+) {
+    let end_comments = comments
         .iter()
         .position(|c| c.start > limit)
         .unwrap_or(comments.len());
-    let popped = comments
-        .get(0..end)
-        .expect("Comments before slicing popped")
+    let end_empty_lines = empty_lines
+        .iter()
+        .position(|l| *l > limit)
+        .unwrap_or(empty_lines.len());
+    let popped_comments = comments
+        .get(0..end_comments)
+        .unwrap()
         .iter()
-        .map(|c| c.content);
+        .map(|c| (c.start, Some(c.content)));
+    let popped_empty_lines: Box<dyn Iterator<Item = _>> = if retain_empty_lines {
```

Comment:
Using a `Box<dyn Iterator>` feels very wrong here, but I couldn't figure out another way to use `std::iter::empty()` and make the call to `merge_by` reusable. I'm still pretty fresh with Rust, so there's probably something cleaner I'm missing 🙂

Reply:
Agreed. I'm not sure what is best here.
Could we move the iterator methods out of the `if else` and have something like this

```rust
if retain_empty_lines {
    empty_lines
} else {
   &[] 
}
    .get(.....) etc
```

If that doesn't work maybe a static slice would help
```rust
static EMPTY_USIZE_SLICE: &'static [usize] = &[];
```